File: src/backend/routes/blockchain.js
const express = require('express');
const router = express.Router();
const ContractEvent = require('../models/contractEvent');
const { authenticate } = require('../middleware/auth');

// Get events by contract name
router.get('/events/:contractName', authenticate, async (req, res) => {
  try {
    const { contractName } = req.params;
    const { page = 1, limit = 20, eventName } = req.query;
    
    const query = { contractName };
    if (eventName) {
      query.eventName = eventName;
    }
    
    const events = await ContractEvent.find(query)
      .sort({ blockNumber: -1, logIndex: -1 })
      .skip((page - 1) * limit)
      .limit(parseInt(limit))
      .lean();
      
    const total = await ContractEvent.countDocuments(query);
    
    res.json({
      success: true,
      data: {
        events,
        pagination: {
          total,
          page: parseInt(page),
          limit: parseInt(limit),
          pages: Math.ceil(total / limit)
        }
      }
    });
  } catch (error) {
    console.error('Error fetching events:', error);
    res.status(500).json({ success: false, message: 'Failed to fetch events' });
  }
});

// Get event by transaction hash
router.get('/events/tx/:txHash', authenticate, async (req, res) => {
  try {
    const { txHash } = req.params;
    
    const events = await ContractEvent.find({ transactionHash: txHash })
      .sort({ logIndex: 1 })
      .lean();
      
    res.json({
      success: true,
      data: events
    });
  } catch (error) {
    console.error('Error fetching event by tx hash:', error);
    res.status(500).json({ success: false, message: 'Failed to fetch event' });
  }
});

// Get events for a specific address (as participant)
router.get('/events/address/:address', authenticate, async (req, res) => {
  try {
    const { address } = req.params;
    const { page = 1, limit = 20 } = req.query;
    
    // This is a complex query that needs to find events where the address is in any of the args
    // For MongoDB, we'll use $or with multiple conditions for different event types
    
    const mentorRegistryEvents = await ContractEvent.find({
      contractName: 'mentorRegistry',
      $or: [
        { 'args.mentorAddress': address.toLowerCase() }
      ]
    })
    .sort({ blockNumber: -1, logIndex: -1 })
    .limit(parseInt(limit))
    .lean();
    
    const sessionManagerEvents = await ContractEvent.find({
      contractName: 'sessionManager',
      $or: [
        { 'args.mentor': address.toLowerCase() },
        { 'args.mentee': address.toLowerCase() }
      ]
    })
    .sort({ blockNumber: -1, logIndex: -1 })
    .limit(parseInt(limit))
    .lean();
    
    const tokenEvents = await ContractEvent.find({
      contractName: 'mentorshipToken',
      $or: [
        { 'args.mentor': address.toLowerCase() },
        { 'args.mentee': address.toLowerCase() },
        { 'args.from': address.toLowerCase() },
        { 'args.to': address.toLowerCase() }
      ]
    })
    .sort({ blockNumber: -1, logIndex: -1 })
    .limit(parseInt(limit))
    .lean();
    
    // Combine all events and sort by blockNumber and logIndex
    const events = [...mentorRegistryEvents, ...sessionManagerEvents, ...tokenEvents]
      .sort((a, b) => {
        if (a.blockNumber !== b.blockNumber) {
          return b.blockNumber - a.blockNumber;
        }
        return b.logIndex - a.logIndex;
      })
      .slice(0, parseInt(limit));
    
    res.json({
      success: true,
      data: {
        events,
        pagination: {
          page: parseInt(page),
          limit: parseInt(limit)
        }
      }
    });
  } catch (error) {
    console.error('Error fetching events for address:', error);
    res.status(500).json({ success: false, message: 'Failed to fetch events' });
  }
});

module.exports = router; 

File: backend/app/routes/blockchain.py
from fastapi import APIRouter, Depends, HTTPException, status, Path, Query
from typing import Dict, Any, List, Optional

from app.utils.auth import get_current_user, TokenData

router = APIRouter(prefix="/blockchain", tags=["Blockchain"])

@router.get("/events/{contract_name}", response_model=Dict[str, Any])
async def get_events_by_contract(
    contract_name: str = Path(...),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    event_name: Optional[str] = Query(None),
    current_user: TokenData = Depends(get_current_user)
):
    """
    Get events by contract name
    """
    # In a real application, you would query your database
    # for contract events based on the contract name
    
    return {
        "success": True,
        "data": {
            "events": [],
            "pagination": {
                "total": 0,
                "page": page,
                "limit": limit,
                "pages": 0
            }
        }
    }

@router.get("/events/tx/{tx_hash}", response_model=Dict[str, Any])
async def get_event_by_tx_hash(
    tx_hash: str = Path(...),
    current_user: TokenData = Depends(get_current_user)
):
    """
    Get event by transaction hash
    """
    # In a real application, you would query your database
    # for contract events based on the transaction hash
    
    return {
        "success": True,
        "data": []
    }

@router.get("/events/address/{address}", response_model=Dict[str, Any])
async def get_events_by_address(
    address: str = Path(...),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    current_user: TokenData = Depends(get_current_user)
):
    """
    Get events for a specific address (as participant)
    """
    # In a real application, you would query your database
    # for contract events where the address is a participant
    
    return {
        "success": True,
        "data": {
            "events": [],
            "pagination": {
                "page": page,
                "limit": limit
            }
        }
    }

@router.get("/tokens", response_model=Dict[str, Any])
async def get_tokens():
    """
    Get token info
    """
    return {
        "success": True,
        "message": "Get tokens route"
    }

@router.get("/mentors", response_model=Dict[str, Any])
async def get_blockchain_mentors():
    """
    Get blockchain mentor list
    """
    return {
        "success": True,
        "message": "Get blockchain mentors route"
    }

@router.post("/transactions", response_model=Dict[str, Any])
async def create_blockchain_transaction(current_user: TokenData = Depends(get_current_user)):
    """
    Create a blockchain transaction
    """
    return {
        "success": True,
        "message": "Create blockchain transaction route"
    } 